src/devices/cpu/drcbearm64.cpp

```diff
@@ -310,11 +310,15 @@ a64::Gp drcbe_arm64::select_register(a64::Gp const &reg, uint32_t regsize) const
 bool drcbe_arm64::is_valid_immediate_mask(uint64_t val, size_t bytes)
 {
 	// all zeros and all ones aren't allowed, and disallow any value with bits outside of the max bit range
-	if (val == 0 || val == make_bitmask<uint64_t>(bytes * 8))
+	if (val == 0 || val == make_bitmask<uint64_t>(bytes * 8) || (bytes == 4 && (val & ~make_bitmask<uint64_t>(bytes * 8)) != 0))
 		return false;
 
-	uint32_t head = 64 - count_leading_zeros_64(val);
-	uint32_t tail = 0;
+	int32_t head = (bytes * 8) - (int32_t)count_leading_zeros_64(val);
+
+	if (head < 0)
+		return false;
+
+	int32_t tail = 0;
```

Comment:
Are you sure the `(head < 0)` check is correct?  Consider:

* Suppose `bytes` = 4, `val` = 0x00000000'00ffff00
* `count_leading_zeros_64(val)` = 40
* `(bytes * 8) - (int32_t)count_leading_zeros_64(val)` = 32 - 40 = -8
* The test will fail if there are no bits outside the low 32 bits set
* The function will never return `true` when `bytes` = 4 because this clashes with the first test

I think what you actually want is more like:
```c++
	// all zeros and all ones aren't allowed, and disallow any value with bits outside of the max bit range
	if (val == 0 || val == make_bitmask<uint64_t>(bytes * 8))
		return false;

	uint32_t head = 64 - count_leading_zeros_64(val);
	if (head >= (bytes * 8))
		return false;

	uint32_t tail = 0;
```

Reply:
Having a very off day today. Yeah that works.